services/core-web/common/models/documents/document.ts

```diff
@@ -0,0 +1,153 @@
+by the table should be *set* to the specific object, cannot expect to be able to consistently call its methods
+
+include "user_roles" property in the json used in the constructor to set allowed actions based on the user
+*/
+export class MineDocument {
+  public mine_document_guid: string;
+
+  public mine_guid: string;
+
+  public document_manager_guid: string;
+
+  public document_name: string;
+
+  public create_user: string;
+
+  public update_user: string;
+
+  public upload_date: string;
+
+  public update_timestamp: string;
+
+  public is_archived: boolean;
+
+  public archived_by: string;
+
+  public archived_date: string;
+
+  public is_latest_version: boolean;
+
+  public category?: string;
+
+  // generated
+  public key: string;
+
+  public file_type: string;
+
+  public number_prev_versions: number;
+
+  public versions: MineDocument[]; // all versions, including latest
+
+  public allowed_actions: FileOperations[];
+
+  constructor(jsonObject: any) {
+    this.mine_document_guid = jsonObject.mine_document_guid;
+    this.mine_guid = jsonObject.mine_guid;
+    this.document_manager_guid = jsonObject.document_manager_guid;
+    this.document_name = jsonObject.document_name;
+    this.create_user = jsonObject.create_user;
+    this.update_user = jsonObject.update_user;
+    this.upload_date = jsonObject.upload_date;
+    this.update_timestamp = jsonObject.update_timestamp;
+    this.category = jsonObject.category;
+    this.is_archived = jsonObject.is_archived ?? false;
+    this.archived_by = jsonObject.archived_by;
+    this.archived_date = jsonObject.archived_date;
+    this.is_latest_version = jsonObject.is_latest_version ?? true;
+    this.setCalculatedProperties(jsonObject);
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  protected makeChild(params: any, _constructorArgs: any) {
+    return new MineDocument(params);
+  }
+
+  protected setCalculatedProperties(jsonObject: any) {
+    this.key = this.is_latest_version
+      ? this.mine_document_guid
+      : jsonObject.document_manager_version_guid;
+    this.file_type = this.getFileType();
+
+    const versions = jsonObject.versions ?? [];
+    if (this.is_latest_version && versions.length) {
+      this.number_prev_versions = versions.length - 1;
+      this.versions = versions
+        .slice(1)
+        .map((version) => this.makeChild({ ...version, is_latest_version: false }, jsonObject));
+    } else {
+      this.number_prev_versions = 0;
+      this.versions = [];
+    }
+    this.setAllowedActions(jsonObject.user_roles);
+  }
+
+  public getFileType() {
+    const index = this.document_name.lastIndexOf(".");
+    return index === -1 ? null : this.document_name.substring(index).toLocaleLowerCase();
+  }
+
+  public setAllowedActions(userRoles: string[] = []) {
+    this.allowed_actions = this.getAllowedActions(userRoles).filter(Boolean);
```

Comment:
https://bcmines.atlassian.net/browse/MDS-5356?focusedCommentId=22159 <-- some explanation of why things are set up this way (with functions setting a value instead of returning it)